# Homework W01 - Introduction

The goal of this homework is to train a simple model for predicting the duration of a ride - similar to what we did in this module.

## Q1. Downloading the Data

We'll use the same NYC taxi dataset, but instead of "Green Taxi Trip Records", we'll use "Yellow Taxi Trip Records".

Download the data for January and February 2023.

Read the data for January. How many columns are there?

* 16
* 17
* 18
* 19

In [1]:
import pandas as pd
taxi_df = pd.read_parquet('../dataset/yellow_tripdata_2023-01.parquet')
print(taxi_df.shape)

(3066766, 19)


Note: dataframe shape returns a tuple showing the dimension of the dataframe. It shows the number of row and column: 3,066,766 rows and 19 columns

## Q2. Computing duration
Now let's compute the `duration` variable. It should contain the duration of a ride in minutes.

What's the standard deviation of the trips duration in January?

* 32.59
* 42.59
* 52.59
* 62.59

In [2]:
taxi_df.head(10)

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         2  2023-01-01 00:32:10   2023-01-01 00:40:36              1.0   
1         2  2023-01-01 00:55:08   2023-01-01 01:01:27              1.0   
2         2  2023-01-01 00:25:04   2023-01-01 00:37:49              1.0   
3         1  2023-01-01 00:03:48   2023-01-01 00:13:25              0.0   
4         2  2023-01-01 00:10:29   2023-01-01 00:21:19              1.0   
5         2  2023-01-01 00:50:34   2023-01-01 01:02:52              1.0   
6         2  2023-01-01 00:09:22   2023-01-01 00:19:49              1.0   
7         2  2023-01-01 00:27:12   2023-01-01 00:49:56              1.0   
8         2  2023-01-01 00:21:44   2023-01-01 00:36:40              1.0   
9         2  2023-01-01 00:39:42   2023-01-01 00:50:36              1.0   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0           0.97         1.0                  N           161           141   
1           1.10         1.0                  N            43           237   
2           2.51         1.0                  N            48           238   
3           1.90         1.0                  N           138             7   
4           1.43         1.0                  N           107            79   
5           1.84         1.0                  N           161           137   
6           1.66         1.0                  N           239           143   
7          11.70         1.0                  N           142           200   
8           2.95         1.0                  N           164           236   
9           3.01         1.0                  N           141           107   

   payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0             2          9.3   1.00      0.5        0.00           0.0   
1             1          7.9   1.00      0.5        4.00           0.0   
2             1         14.9   1.00      0.5       15.00           0.0   
3             1         12.1   7.25      0.5        0.00           0.0   
4             1         11.4   1.00      0.5        3.28           0.0   
5             1         12.8   1.00      0.5       10.00           0.0   
6             1         12.1   1.00      0.5        3.42           0.0   
7             1         45.7   1.00      0.5       10.74           3.0   
8             1         17.7   1.00      0.5        5.68           0.0   
9             2         14.9   1.00      0.5        0.00           0.0   

   improvement_surcharge  total_amount  congestion_surcharge  airport_fee  
0                    1.0         14.30                   2.5         0.00  
1                    1.0         16.90                   2.5         0.00  
2                    1.0         34.90                   2.5         0.00  
3                    1.0         20.85                   0.0         1.25  
4                    1.0         19.68                   2.5         0.00  
5                    1.0         27.80                   2.5         0.00  
6                    1.0         20.52                   2.5         0.00  
7                    1.0         64.44                   2.5         0.00  
8                    1.0         28.38                   2.5         0.00  
9                    1.0         19.90                   2.5         0.00

In [3]:
taxi_df['duration'] = taxi_df.tpep_dropoff_datetime - taxi_df.tpep_pickup_datetime
taxi_df['duration'] = taxi_df.duration.dt.total_seconds() / 60
print("Standard Deviation for duration data: ", taxi_df['duration'].std())

Standard Deviation for duration data:  42.59435124195458


## Q3. Dropping outliers
Next, we need to check the distribution of the `duration` variable. There are some outliers. Let's remove them and keep only the records where the duration was between 1 and 60 minutes (inclusive).

What fraction of the records left after you dropped the outliers?

* 90%
* 92%
* 95%
* 98%

In [4]:
df = taxi_df[(taxi_df.duration >= 1) & (taxi_df.duration <= 60)]
print("records where the duration is between 1 and 60 minutes: ", df.shape)

records where the duration is between 1 and 60 minutes:  (3009173, 20)


In [5]:
percentage = (df.shape[0])/taxi_df.shape[0]
print("records left after dropping outliers: " ,f"{percentage:.2%}")

records left after dropping outliers:  98.12%


In [6]:
df.head(5)

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         2  2023-01-01 00:32:10   2023-01-01 00:40:36              1.0   
1         2  2023-01-01 00:55:08   2023-01-01 01:01:27              1.0   
2         2  2023-01-01 00:25:04   2023-01-01 00:37:49              1.0   
3         1  2023-01-01 00:03:48   2023-01-01 00:13:25              0.0   
4         2  2023-01-01 00:10:29   2023-01-01 00:21:19              1.0   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0           0.97         1.0                  N           161           141   
1           1.10         1.0                  N            43           237   
2           2.51         1.0                  N            48           238   
3           1.90         1.0                  N           138             7   
4           1.43         1.0                  N           107            79   

   payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0             2          9.3   1.00      0.5        0.00           0.0   
1             1          7.9   1.00      0.5        4.00           0.0   
2             1         14.9   1.00      0.5       15.00           0.0   
3             1         12.1   7.25      0.5        0.00           0.0   
4             1         11.4   1.00      0.5        3.28           0.0   

   improvement_surcharge  total_amount  congestion_surcharge  airport_fee  \
0                    1.0         14.30                   2.5         0.00   
1                    1.0         16.90                   2.5         0.00   
2                    1.0         34.90                   2.5         0.00   
3                    1.0         20.85                   0.0         1.25   
4                    1.0         19.68                   2.5         0.00   

    duration  
0   8.433333  
1   6.316667  
2  12.750000  
3   9.616667  
4  10.833333

## Q4. One-hot encoding
Let's apply one-hot encoding to the pickup and dropoff location IDs. We'll use only these two features for our model.

Turn the dataframe into a list of dictionaries (remember to re-cast the ids to strings - otherwise it will label encode them)
* Fit a dictionary vectorizer
* Get a feature matrix from it

What's the dimensionality of this matrix (number of columns)?

* 2
* 155
* 345
* 515
* 715

In [7]:
print(df['PULocationID'].isnull().sum())
print(df['DOLocationID'].isnull().sum())

0
0


Note: no `null` values in `PULocationID` and `DOLocationID` variables

In [8]:
from sklearn.feature_extraction import DictVectorizer

cat = ['PULocationID', 'DOLocationID']
train_dicts = df[cat].astype('int').astype('str').to_dict(orient='records')
dv = DictVectorizer()
X_train = dv.fit_transform(train_dicts)
X_train.shape

(3009173, 515)

Note: dataframe shape returns a tuple showing the dimension of the dataframe. It shows the number of row and column: 3,009,173 rows and 515 columns

## Q5. Training a model
Now let's use the feature matrix from the previous step to train a model.

* Train a plain linear regression model with default parameters
* Calculate the RMSE of the model on the training data

What's the RMSE on train?

* 3.64
* 7.64
* 11.64
* 16.64

In [9]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import root_mean_squared_error

y_train = df.duration.values

lr = LinearRegression()
lr.fit(X_train, y_train)
y_pred = lr.predict(X_train)

rmse_train  = root_mean_squared_error(y_train, y_pred)

print(rmse_train)

7.649261932106969


## Q6. Evaluating the model
Now let's apply this model to the validation dataset (February 2023).

What's the RMSE on validation?

* 3.81
* 7.81
* 11.81
* 16.81

In [10]:
categorical = ['PULocationID', 'DOLocationID']

def read_data(filename):
    df = pd.read_parquet(filename)
    
    df['duration'] = df.tpep_dropoff_datetime - df.tpep_pickup_datetime
    df['duration'] = df.duration.dt.total_seconds() / 60

    df = df[(df.duration >= 1) & (df.duration <= 60)].copy()

    df[categorical] = df[categorical].astype('int').astype('str')
    
    return df

In [11]:
df_val = read_data('../dataset/yellow_tripdata_2023-02.parquet')
df_val.head(5)

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         1  2023-02-01 00:32:53   2023-02-01 00:34:34              2.0   
3         1  2023-02-01 00:29:33   2023-02-01 01:01:38              0.0   
4         2  2023-02-01 00:12:28   2023-02-01 00:25:46              1.0   
5         1  2023-02-01 00:52:40   2023-02-01 01:07:18              1.0   
6         1  2023-02-01 00:12:39   2023-02-01 00:40:36              1.0   

   trip_distance  RatecodeID store_and_fwd_flag PULocationID DOLocationID  \
0           0.30         1.0                  N          142          163   
3          18.80         1.0                  N          132           26   
4           3.22         1.0                  N          161          145   
5           5.10         1.0                  N          148          236   
6           8.90         1.0                  N          137          244   

   payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0             2          4.4   3.50      0.5        0.00           0.0   
3             1         70.9   2.25      0.5        0.00           0.0   
4             1         17.0   1.00      0.5        3.30           0.0   
5             1         21.9   3.50      0.5        5.35           0.0   
6             1         41.5   3.50      0.5        3.50           0.0   

   improvement_surcharge  total_amount  congestion_surcharge  Airport_fee  \
0                    1.0          9.40                   2.5         0.00   
3                    1.0         74.65                   0.0         1.25   
4                    1.0         25.30                   2.5         0.00   
5                    1.0         32.25                   2.5         0.00   
6                    1.0         50.00                   2.5         0.00   

    duration  
0   1.683333  
3  32.083333  
4  13.300000  
5  14.633333  
6  27.950000

In [12]:
val_dicts = df_val[categorical].to_dict(orient='records')
X_val = dv.transform(val_dicts)
X_val.shape

(2855951, 515)

In [13]:
y_pred = lr.predict(X_val)
y_val = df_val.duration.values
rmse_val = root_mean_squared_error(y_val, y_pred)
print(rmse_val)

7.811818743246608
